protocol/x/ratelimit/util/capacity.go

```diff
@@ -0,0 +1,96 @@
+package util
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/dydxprotocol/v4-chain/protocol/dtypes"
+	"github.com/dydxprotocol/v4-chain/protocol/lib"
+	"github.com/dydxprotocol/v4-chain/protocol/x/ratelimit/types"
+)
+
+// CalculateNewCapacityList calculates the new capacity list for the given current `tvl` and `limitParams“.
+// Input invariant: `len(prevCapapcityList) == len(limitParams.Limiters)`
+// Detailed math for calculating the updated capacity:
+//
+//	`baseline = max(baseline_minimum, baseline_tvl_ppm * tvl)`
+//	`capacity_diff = max(baseline, capacity-baseline) * (time_since_last_block / period)`
+//
+// This is basically saying that the capacity returns to the baseline over the course of the `period`.
+// Usually in a linear way, but if the `capacity` is more than twice the `baseline`, then in an exponential way.
+//
+//	`capacity =`
+//	    if `abs(capacity - baseline) < capacity_diff` then `capacity = baseline`
+//	    else if `capacity < baseline` then `capacity += capacity_diff`
+//	    else `capacity -= capacity_diff`
+//
+// On a high level, `capacity` trends towards `baseline` by `capacity_diff` but does not “cross” it.
+func CalculateNewCapacityList(
+	bigTvl *big.Int,
+	limitParams types.LimitParams,
+	prevCapapcityList []dtypes.SerializableInt,
+	timeSinceLastBlock time.Duration,
+) (newCapacityList []dtypes.SerializableInt) {
+	// Declare new capacity list to be populated.
+	newCapacityList = make([]dtypes.SerializableInt, len(prevCapapcityList))
+
+	for i, limiter := range limitParams.Limiters {
+		// For each limiter, calculate the current baseline.
+		baseline := GetBaseline(bigTvl, limiter)
+
+		capacityMinusBaseline := new(big.Int).Sub(
+			prevCapapcityList[i].BigInt(), // array access is safe because of input invariant
+			baseline,
+		)
+
+		// Calculate left operand: `max(baseline, capacity-baseline)`. This equals `baseline` when `capacity <= 2 * baseline`
+		operandL := new(big.Rat).SetInt(
+			lib.BigMax(
+				baseline,
+				capacityMinusBaseline,
+			),
+		)
+
+		// Calculate right operand: `time_since_last_block / period`
+		periodMilli := new(big.Int).Mul(
+			new(big.Int).SetUint64(uint64(limiter.PeriodSec)),
+			big.NewInt(1000),
+		)
+		operandR := new(big.Rat).SetFrac(
+			new(big.Int).SetInt64(timeSinceLastBlock.Milliseconds()),
+			periodMilli,
+		)
```

Comment:
nit: if we do use the `time.Duration` type (see above comment) then this can just use `SetFrac64`